Add the Viewer CRD controller for managing web views such as Tensorboard instances from within the Pipelines UI.

neuromage opened the pull request:
This PR adds a basic controller for managing a new type of CRD for managing instances of web views such as Tensorboard from within the Pipelines system.

Currently, the controller only supports Tensorboard views, but additional views should be trivial to add.

This PR makes use of the new Kubernetes controller-runtime libraries, which is the set of libraries underlying kubebuilder. These libraries make creating CRD controllers a lot easier by managing internal data structures like informers and workqueues, so we can focus on writing just the business logic.

I still need to do the following before merging this PR:

- [ ] Add tests
- [ ] Update the README to describe the approach and how to run this controller
- [ ] Add YAML definitions for the Viewer CRD type

For now, I'd like to get some feedback on this approach, before adding any tests.

Once this PR goes in, I will send a separate PR for bundling this controller and deploying it as part of pipelines.

neuromage commented:
Fixes #443

neuromage commented:
/retest

vicaire commented on backend/src/crd/controller/viewer/reconciler/reconciler.go:
```diff
@@ -0,0 +1,224 @@
+	return reconcile.Result{}, nil
+}
+
+func setPodSpecForTensorboard(view *viewerV1alpha1.Viewer, s *corev1.PodSpec) {
```
Are you setting the owner references for the created resources so that garbage collection is taken care of? 

Do you need labels so that we can query by type of resource? Otherwise the UI may list all the resources which could be a problem once we add additional types.

vicaire commented on backend/src/crd/controller/viewer/reconciler/reconciler.go:
```diff
@@ -0,0 +1,224 @@
+		}
+	}
+	glog.Infof("Created new deployment with spec: %+v", dpl)
+
```
Is there a way to write simple fakes like in the case of the persistence agent: 

https://github.com/kubeflow/pipelines/tree/master/backend/src/agent/persistence/client

So that we can test end to end?

neuromage replied:
Yes, there is. This would be what we'd use:
https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/reconcile/reconciletest

I will add an e2e test in a follow up PR. For now, I added unit tests testing the behaviour of the reconciler using the fake kubernetes client here:
https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/client/fake

neuromage commented on backend/src/crd/pkg/apis/viewer/register.go:
```diff
@@ -0,0 +1,25 @@
+package viewer
+
+const (
+	Kind      string = "Viewer"
```
Yeah, I agree, it's not a great name. I thought of viewer since it's meant to proxy views of another webapp through ambassador that we launch and control. I can't think of another one right now, and I'm open to suggestions :-)

neuromage commented:
Added unit tests and CRD definitions. This is ready to be looked at again.

/retest

neuromage commented:
/retest

neuromage commented:
/assign @IronPan

yebrahim commented on backend/src/crd/pkg/apis/viewer/v1alpha1/types.go:
```diff
@@ -0,0 +1,77 @@
+type TensorboardSpec struct {
+	// LogDir is the location of the log directory to be read by tensorboard, i.e.,
+	// ---log_dir.
+	LogDir string `json:"logDir"`
```
This should be an array (slice?) of strings. Tensorboard supports comparing outputs by passing a list of logdirs, we're using this feature in run comparison.

neuromage replied:
It's just a comma separated list right? Which can be passed in directly here. This may be more flexible, in that I'm not reinterpreting the arguments to Tensorboard. For example, TB allows naming some of those runs as well. Rather than structure that here, just pass in the expected format to Tensorboard. What do you think?

yebrahim replied:
That's fine I guess, although structuring it would give some value, because the structure is not very straightforward. Something like "name1:<encoded_path_1>,name2:<encoded_path_2>" is easy to mistype.

neuromage replied:
True, but if we're planning on creating the viewer through UI, then we can try to construct it for the user there instead before sending the create request?

yebrahim replied:
Yup, we can have clients do it.

yebrahim commented on backend/src/crd/pkg/apis/viewer/v1alpha1/types.go:
```diff
@@ -0,0 +1,77 @@
+const (
+	// ViewerTypeTensorboard is the ViewerType constant used to indicate that the
+	// underlying type is Tensorboard. An instance named `instance123` will serve
+	// under /tensorboard/instance123.
```
We should also consider mapping logdirs to instances in the route if we can, so that the user can just navigate to /tensorboard/encoded-paths-here, and they get routed to a new or existing instance by the CRD. Is this doable?

neuromage replied:
I looked into this a bit, and at this point, I don't think we can. In that, the CRD isn't watching the routes, so it can't create a new one when the user navigates to it. The other thing is, I think it's worth having short names (right now, auto-generated by k8s) to give a shorter user-facing url. Mapping the logdirs to a name may result in a longer string, which I'm not in favour of.

Tangential to all this is keeping a unique a instance of TB for a given set of logdirs. I think this is a nice to have optimization, and I'm happy to look into adding this functionality in a follow up PR. What do you think?

yebrahim replied:
Who keeps track of that mapping though?

This isn't P0 by any means for this implementation, but it's good to keep it in mind while we're working on it.

neuromage replied:
I imagine the right approach would be to label the viewer with the logdirs, which you can select for to ensure it's not already created. The reconciler here aims to be stateless as much as possible.

yebrahim replied:
SGTM.

yebrahim commented:
/lgtm

IronPan commented on backend/src/crd/controller/viewer/reconciler/reconciler.go:
```diff
@@ -0,0 +1,274 @@
+// Package reconciler describes a Reconciler for working with Viewer CRDs. The
+// reconciler takes care of the main logic for ensuring every Viewer CRD
+// corresponds to a unique deployment and backing service. The service is
+// annotated such that it is compatible with Ambassador managed routing.
```
is it a requirement to have ambassador as dependency? Can we decouple pipeline from ambassador?

neuromage replied:
As discussed offline, I removed explicit references to Ambassador. This now works both with and without (via direct port forwarding). I added instructions in the README for how to use the latter approach as well.

IronPan commented on backend/src/crd/controller/viewer/reconciler/reconciler.go:
```diff
@@ -0,0 +1,274 @@
+	s.Containers = append(s.Containers, c)
+}
+
+func deploymentFrom(view *viewerV1alpha1.Viewer) (*appsv1.Deployment, error) {
```
is it possible to launch a TB pod with a volume mounted, instead of a gcs path?
I guess it's as simple as passing right spec to PodTemplateSpec. is that correct?

neuromage replied:
Pretty much. I did have to make a slight change to how I set up the container so I can reuse the existing volume mount if one is specified. I also added a unit test for this, and verified it works in my own cluster. I will add testing this functionality as part of my e2e test in an upcoming PR.

IronPan commented:
Do you mind having a test/sample about using local logdir with persistent volume?

neuromage commented:
Added tests and sample for persistent volumes, and also removed the explicit dependency on Ambassador. Routing should work through direct port forwarding now as well, and I added this to the README. PTAL.

neuromage commented:
/retest

neuromage commented:
/retest

IronPan commented:
/test kubeflow-pipeline-e2e-test

IronPan commented:
/lgtm
/approve
